florestad/src/json_rpc/blockchain.rs

```diff
@@ -44,6 +44,30 @@ impl<Blockchain: RpcChain> RpcImpl<Blockchain> {
             .get_block(&blockhash)
             .map_err(|_| JsonRpcError::BlockNotFound)
     }
+
+    /// Retrieves the height of the block that was mined in the given(or less) timestamp.
+    /// Intended to use for rescan requests that only has timestamp as info.
+    ///
+    /// This function uses the blockchain principle that a block is mined in around 10 minutes
+    /// to fetch a blockheight.
+    ///
+    /// The main formula to start searching for the block is (timestamp - GENESIS_TIMESTAMP) / 600
+    ///
+    /// Inserting an invalid timestamp will result in returning Ok(0u32).
+    pub async fn get_block_height_by_timestamp(&self, timestamp: u32) -> Result<u32, JsonRpcError> {
```

Comment:
You don't need to download the block here, we have the header store in `ChainState`